src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs

```diff
@@ -0,0 +1,506 @@
+                      null,
+                      null,
+                      s_useMachineKeyStore))
+        {
+        }
+
+        public DSACryptoServiceProvider(int dwKeySize)
+            : this(dwKeySize,
+                  new CspParameters(CapiHelper.DefaultDssProviderType,
+                      null,
+                      null,
+                      s_useMachineKeyStore))
+        {
+        }
+
+        public DSACryptoServiceProvider(CspParameters parameters)
+            : this(0, parameters)
+        {
+        }
+
+        public DSACryptoServiceProvider(int dwKeySize, CspParameters parameters)
+        {
+            if (dwKeySize < 0)
+                throw new ArgumentOutOfRangeException(nameof(dwKeySize), SR.ArgumentOutOfRange_NeedNonNegNum);
+
+            _parameters = CapiHelper.SaveCspParameters(
+                CapiHelper.CspAlgorithmType.Dss,
+                parameters,
+                s_useMachineKeyStore,
+                out _randomKeyContainer);
+
+            _keySize = dwKeySize;
+            _sha1 = SHA1.Create();
+
+            // If this is not a random container we generate, create it eagerly 
+            // in the constructor so we can report any errors now.
+            if (!_randomKeyContainer)
+            {
+                // Force-read the SafeKeyHandle property, which will summon it into existence.
+                SafeKeyHandle localHandle = SafeKeyHandle;
+                Debug.Assert(localHandle != null);
+            }
+        }
+
+        private SafeProvHandle SafeProvHandle
+        {
+            get
+            {
+                if (_safeProvHandle == null)
+                {
+                    lock (_parameters)
+                    {
+                        if (_safeProvHandle == null)
+                        {
+                            SafeProvHandle hProv = CapiHelper.CreateProvHandle(_parameters, _randomKeyContainer);
+
+                            Debug.Assert(hProv != null);
+                            Debug.Assert(!hProv.IsInvalid);
+                            Debug.Assert(!hProv.IsClosed);
+
+                            _safeProvHandle = hProv;
+                        }
+                    }
+
+                    return _safeProvHandle;
+                }
+
+                return _safeProvHandle;
+            }
+            set
+            {
+                lock (_parameters)
+                {
+                    SafeProvHandle current = _safeProvHandle;
+
+                    if (ReferenceEquals(value, current))
+                    {
+                        return;
+                    }
+
+                    if (current != null)
+                    {
+                        SafeKeyHandle keyHandle = _safeKeyHandle;
+                        _safeKeyHandle = null;
+                        keyHandle?.Dispose();
+                    }
+
+                    _safeProvHandle = value;
+                    current?.Dispose();
```

Comment:
Nit: should this be done in the current != null block, such that you could then just use ```current.Dispose()```?